File: backend/providers/anthropic_provider.py
import base64
from typing import AsyncIterator, Optional

from anthropic import AsyncAnthropic

from .base import LLMProvider


class AnthropicProvider(LLMProvider):
    name = "anthropic"

    def __init__(self, default_model: str = "claude-sonnet-4-6"):
        self.default_model = default_model

    async def stream_vision_response(
        self,
        image_data_url: str,
        transcript_context: str,
        system_prompt: str,
        api_key: Optional[str] = None,
    ) -> AsyncIterator[str]:
        client = AsyncAnthropic(api_key=api_key)

        header, b64data = image_data_url.split(",", 1)
        media_type = header.split(";")[0].split(":")[1]  # e.g. image/png

        user_text = (
            f"Recent conversation transcript:\n{transcript_context}\n\n"
            "Here is the current screen. Respond with the most useful, "
            "concise assistance for what's happening right now."
            if transcript_context
            else "Here is the current screen. Respond with the most useful, "
            "concise assistance for what's happening right now."
        )

        async with client.messages.stream(
            model=self.default_model,
            max_tokens=1024,
            system=system_prompt,
            messages=[
                {
                    "role": "user",
                    "content": [
                        {
                            "type": "image",
                            "source": {
                                "type": "base64",
                                "media_type": media_type,
                                "data": b64data,
                            },
                        },
                        {"type": "text", "text": user_text},
                    ],
                }
            ],
        ) as stream:
            async for text in stream.text_stream:
                yield text

File: android/app/src/main/java/com/wisp/app/providers/AnthropicProvider.kt
package com.wisp.app.providers

import kotlinx.coroutines.flow.Flow
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject

class AnthropicProvider(private val model: String = "claude-sonnet-4-6") : LLMProvider {
    override fun streamVisionResponse(
        imageBase64: String,
        mimeType: String,
        transcriptContext: String,
        systemPrompt: String,
        apiKey: String,
    ): Flow<String> {
        val content = JSONArray()
            .put(
                JSONObject()
                    .put("type", "image")
                    .put(
                        "source",
                        JSONObject()
                            .put("type", "base64")
                            .put("media_type", mimeType)
                            .put("data", imageBase64),
                    ),
            )
            .put(JSONObject().put("type", "text").put("text", buildUserText(transcriptContext)))

        val body = JSONObject()
            .put("model", model)
            .put("max_tokens", 1024)
            .put("system", systemPrompt)
            .put("stream", true)
            .put(
                "messages",
                JSONArray().put(JSONObject().put("role", "user").put("content", content)),
            )

        val request = Request.Builder()
            .url("https://api.anthropic.com/v1/messages")
            .addHeader("x-api-key", apiKey)
            .addHeader("anthropic-version", "2023-06-01")
            .post(body.toString().toRequestBody("application/json".toMediaType()))
            .build()

        return streamSse(request) { payload ->
            val json = JSONObject(payload)
            if (json.optString("type") != "content_block_delta") return@streamSse null
            val delta = json.optJSONObject("delta") ?: return@streamSse null
            if (delta.optString("type") != "text_delta") return@streamSse null
            delta.optString("text").takeIf { it.isNotEmpty() }
        }
    }
}
